analyzer/evmverifier/sourcify/client.go

```diff
@@ -0,0 +1,210 @@
+	if err != nil {
+		return nil, fmt.Errorf("sourcify API call failure: %w", err)
+	}
+	defer res.Body.Close()
+	resp, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	s.logger.Debug("sourcify API call response", "method", method, "url", url.String(), "status", res.Status, "response_body", string(resp))
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("sourcify API call failure: %s", res.Status)
+	}
+
+	return resp, nil
+}
+
+// GetVerifiedContractAddresses returns a list of all verified contract addresses for the given runtime.
+//
+// API docs: https://docs.sourcify.dev/docs/api/server/get-contract-addresses-all/
+//
+// Note: This uses the free, public server API. If it turns out to be unreliable, we could use the repository API (vis IPFS proxy) instead, e.g.:
+// http://ipfs.default:8080/ipns/repo.sourcify.dev/contracts/full_match/23294
+func (s *SourcifyClient) GetVerifiedContractAddresses(ctx context.Context, runtime common.Runtime) ([]ethCommon.Address, error) {
+	// Fetch verified contract addresses.
+	u := *s.serverUrl
+	u.Path = path.Join(u.Path, "files/contracts", sourcifyChains[s.chain][runtime])
+	body, err := s.callAPI(ctx, http.MethodGet, &u)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch verified contract addresses: %w", err)
+	}
+
+	// Parse response.
+	var response struct {
+		Full []ethCommon.Address `json:"full"`
+		// XXX: Skip partial matches for now: https://docs.sourcify.dev/docs/full-vs-partial-match/
+	}
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nil, fmt.Errorf("failed to parse verified contract addresses: %w", err)
+	}
+
+	return response.Full, nil
+}
+
+// Metadata represents a subset of the metadata.json file generated by the Solidity compiler.
+// The exact structure of the metadata output can vary depending on the Solidity compiler version,
+// and there is no standard format. However, the fields included in this struct should generally present
+// across different compiler versions and contracts.
+type Metadata struct {
+	Compiler struct {
+		Version string `json:"version"`
+	} `json:"compiler"`
+	Language string          `json:"language"`
+	Settings json.RawMessage `json:"settings"`
+	Output   struct {
+		ABI     abi.ABI         `json:"abi"`
+		Devdoc  json.RawMessage `json:"devdoc"`
+		Userdoc json.RawMessage `json:"userdoc"`
+	} `json:"output"`
+	Version int `json:"version"`
+}
+
+// GetContractMetadataJSON returns the metadata.json for the given contract address that is verified on sourcify.
+//
+// API docs: https://docs.sourcify.dev/docs/api/repository/get-file-static/
+//
+// Note: this could also use the repository API (via IPFS proxy) instead:
+// - https://docs.sourcify.dev/docs/api/repository/get-file-repository/
+func (s *SourcifyClient) GetContractMetadata(ctx context.Context, runtime common.Runtime, address ethCommon.Address) (*Metadata, error) {
```

Comment:
Initially, I didn't have the special handling of metadata in  `GetContractSourceFiles`. 

I kept this in, because I am not sure yet if we plan to also store the source code or not? (if we wouldn't store the source code, then we would just need to fetch the metadata).

The source code can get quite big, but if we store it as an opaque blob, I don't think it would be problematic anytime soon.